### PR TITLE
Issue 1280 - Add feature which generates a github actions file to bui…

### LINKED ISSUE #1280: Add feature which generates a github actions file to build projects
Add a new feature `github-workflow-ci`

# Gradle

see: https://github.com/gradle/gradle-build-action

In github when you go to actions and click `New Workflow`:

Screenshot 2022-06-03 at 12 22 12

We could generate something like: 

```yaml
name: Java CI with Gradle

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

permissions:
  contents: read

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v3
    - name: Set up JDK 11
      uses: actions/setup-java@v3
      with:
        java-version: '11'
        distribution: 'temurin'
     - name: Gradle Build Action
        uses: gradle/gradle-build-action@v2.1.7
      with:
        arguments: build
```

# Maven 

Screenshot 2022-06-03 at 12 23 33

```yaml
name: Java CI with Maven

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v3
    - name: Set up JDK 11
      uses: actions/setup-java@v3
      with:
        java-version: '11'
        distribution: 'temurin'
        cache: maven
    - name: Build with Maven
      run: ./mvnw package
```

### PR DESCRIPTION
**Pull request by wetted**
Closes #1280

### CODE REVIEW

**Comment by wetted** [comment]
Notes and concerns:

- @melix and @alvarosanchez If you have any feedback/suggestions on the contents of the `gradle.yml`/`maven.yml` actions I appreciate any corrections/improvements
- I'm not entirely convinced by adding `JdkDistribution` to the options package. It might come in useful for github actions and Gradle javaToolChain support, so I am open to what others feel about it. The `setup-java` action step requires a dist. For `JdkDistribution` I included everything listed on my M1 Mac for `sdk list java` but only a subset of that is even supported by the `setup-java` step. I believe a complete set of dists might be desirable for Gradle javaToolChains support.
- There is already a starter-core API for other github actions. However, we also want to support some other CI toolchains (Google Cloud Build, AWS CodeBuild, for example). I created a different package for that reason, but we might want to consider a sensible way to unify them (or not).

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+@if (buildTool.equals(BuildTool.MAVEN)) {
+permissions:
+  contents: read
```
Comment (alvarosanchez) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw:23]: Out of curiosity: why is this needed?

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    - uses: actions/checkout@@v3
+    - name: Set up JDK @jdkVersion.majorVersion()
+      uses: actions/setup-java@@v3
+      with:
+        java-version: @jdkVersion.majorVersion()
+        distribution: @jdkDistribution.distribution()
+@if (buildTool.equals(BuildTool.MAVEN)) {
+        cache: maven
+    - name: Build with Maven
+      run: mvn -B package --file pom.xml
```
Comment (alvarosanchez) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw]:
```suggestion
      run: mvn -B verify
```

**Review by alvarosanchez — CHANGES_REQUESTED** [review]

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      uses: gradle/wrapper-validation-action@@v1.0.4
+    - name: Build with Gradle
+      uses: gradle/gradle-build-action@@v2.2.0
+      with:
+        arguments: build
+    - uses: actions/upload-artifact@@v3.1.0
+      with:
+        name: Build Artifacts
+        path: |
+          **/build/libs
```
Comment (melix) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw]: I'm not sure we should upload libraries. Test reports are fine, but libraries sound a bit odd, especially without metadata like when published in a repository.

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw (hunk 1 above)
Comment (wetted) [reply starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw:23]: I have a mistake with that (it should be for Gradle, not Maven). Github actions generates that for the `Java with Gradle` action but not the `Java with Maven` action, but I don't why it's included for Gradle. Maybe @melix knows?

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Override
+    @NonNull
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public String getTitle() {
+        return "Build and test a Java project using a wrapper script";
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java]: Replace with `Github Actions CI Workflow`

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    @NonNull
+    public String getDescription() {
+        return "Adds a GitHub Actions Workflow to build and test a Java project";
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java]: replace `build and test a Java project` to `build a Micronaut application`

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Override
+    public String getThirdPartyDocumentation() {
+        return "https://docs.github.com/en/actions";
+    }
+
+    @Override
+    public String getWorkflowFileName(GeneratorContext generatorContext) {
+        switch (generatorContext.getBuildTool()) {
+            case GRADLE: case GRADLE_KOTLIN:
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java]: break line: 

```
case GRADLE:
case GRADLE_KOTLIN:
```

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    public String getThirdPartyDocumentation() {
+        return "https://docs.github.com/en/actions";
+    }
+
+    @Override
+    public String getWorkflowFileName(GeneratorContext generatorContext) {
+        switch (generatorContext.getBuildTool()) {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java:76]: You could replace: 

```java
        switch (generatorContext.getBuildTool()) {
                    case GRADLE: case GRADLE_KOTLIN:
          return "gradle.yml";
             case MAVEN:
                 return "maven.yml";
             default:
                 throw new IllegalArgumentException("Unexpected constant for BuildTool enum");
         }
```         

with: 

```java
return generatorContext.getBuildTool().isGradle() ? "gradle.yml" : "maven.yml";
```

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/CIWorkflowFeature.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    public boolean supports(ApplicationType applicationType) {
+        return true;
+    }
+
+    @Override
+    public String getCategory() {
+        return Category.CICD;
+    }
+
+    public abstract String getWorkflowFileName(GeneratorContext generatorContext);
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/CIWorkflowFeature.java:40]: add nullablity annotations. Maybe: 

```java
@NonNull
    public abstract String getWorkflowFileName(@NonNull GeneratorContext generatorContext);
```

File: starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.starter.options;
+
+public enum JdkDistribution {
+    ADOPT("adopt", "Adopt OpenJDK Hotspot"),
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java:27]: where did you optain this list from? From github? Can you add a javadoc with a link?

File: starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    private final String distribution;
+    private final String description;
+
+    JdkDistribution(String distribution, String description) {
+        this.distribution = distribution;
+        this.description = description;
+    }
+
+    public String distribution() {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java:47]: Annotate return with `@NonNull`

File: starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        return distribution;
+    }
+
+    public String description() {
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java:52]: Annotate return with `@NonNull`

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java (hunk 7 above)
Comment (wetted) [reply starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java:76]: The contributor's guide – `CONTRIBUTING.md` – under "Style Considerations" recommends avoiding default/empty `else` constructs, which why I didn't use the conditional operator here. I'm fine with whatever you prefer though.

File: starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java (hunk 9 above)
Comment (wetted) [reply starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java:27]: I got that list from SDKMAN on my M1 Mac. I just did a list `sdk list java` but it only gives me the ones for Apple Silicon. If there is an official list of distributions somewhere I'm not aware of it. However, I just found a list at the [SDKMAN website](https://sdkman.io/jdks), so I wonder which of these we do or don't want to enumerate (or all)?

Comment (sdelamo) [reply starter-core/src/main/java/io/micronaut/starter/options/JdkDistribution.java:27]: Update to reflect the distributions supported by github actions `setup-java`

see: https://github.com/actions/setup-java#supported-distributions

add javadoc `@see` with the above link to this enum.

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+JdkVersion jdkVersion,
+JdkDistribution jdkDistribution,
+BuildTool buildTool
+)
+
+name: Java CI with @buildTool.getTitle()
+
+on:
+  push:
+    branches: [ main ]
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/templates/javaAction.rocker.raw]: add a template parameter. `String branch`.

File: starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @Override
+    public void apply(GeneratorContext generatorContext) {
+        super.apply(generatorContext);
+
+        String workflowFilePath = ".github/workflows/" + getWorkflowFileName(generatorContext);
+        generatorContext.addTemplate("javaAction",
+                new RockerTemplate(workflowFilePath,
+                        javaAction.template(
```
Comment (sdelamo) [inline starter-core/src/main/java/io/micronaut/starter/feature/ci/workflows/github/GithubCiWorkflowFeature.java]: extract to method. e.g. `RockerModel workflowFileRockerModel(GeneratorContext generatorContext)`

**Review by sdelamo — CHANGES_REQUESTED** [review]

**Comment by sdelamo** [comment]
Failure relates to https://github.com/micronaut-projects/micronaut-starter/pull/1291

**Comment by sdelamo** [comment]
can you merge 3.6.x into this branch @wetted it should get the build green

**Review by sdelamo — APPROVED** [review]
